### ISSUE TITLE
[QUERY] Monitoring ServiceBusProcessor

### ISSUE THREAD

**Issue by cjablonski76** [Azure/azure-sdk-for-net#44782]
### Library name and version

Azure.Messaging.ServiceBus 7.17.5

### Query/Question

I am using the `ServiceBusProcessor` class to handle processing messages in the background of our asp.net application. Is there a way for us to monitor and alert whether the processors are still healthy and attempting to receive messages?

I have seen in logs messages like:
```
info: Azure.Messaging.ServiceBus[7] => SpanId:89251b9b75bfa1a2, TraceId:8b65abf017c608b5d3a4640e9b0a4310, ParentId:0000000000000000 Name-Receiver: ReceiveBatchAsync start. MessageCount = 1
```
Are logged, but is there any similar metric emitted so that we can effectively monitor our receivers to ensure that there's always one or more receivers available?

### Environment

```
dotnet --info

global.json file:
  Not found

Host:
  Version:      6.0.31
  Architecture: x64
  Commit:       e2ca2f8a1c

.NET SDKs installed:
  No SDKs were found.

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.31 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.31 [/usr/share/dotnet/shared/Microsoft.NETCore.App]

Download .NET:
  https://aka.ms/dotnet-download

Learn about .NET Runtimes and SDKs:
  https://aka.ms/dotnet/runtimes-sdk-info
  ```

**Comment by jsquire** [comment]
Hi @cjablonski76.  Thanks for reaching out and for your question.  The Service Bus processor is designed to be highly resilient to errors and will automatically recover from transient errors and continue processing.  It does not automatically stop under any normal circumstances unless your process crashes.   So long as [IsProcessing](https://learn.microsoft.com/dotnet/api/azure.messaging.servicebus.servicebusprocessor.isprocessing?view=azure-dotnet#azure-messaging-servicebus-servicebusprocessor-isprocessing) returns `true`, the processor is running and actively attempting to receive messages.  Any errors that it encounters, whether part of the internal infrastructure or in your processing handler, are surfaced to the error handler.   If the flag is set and your error handler is not being pinged, then the processor is running, healthy, and looking for messages.

As it is working, the processor emits tracing data that follows the [Open Telemetry Messaging specification](https://opentelemetry.io/docs/specs/semconv/attributes-registry/messaging/), which can be collected using any OTel tooling, such as the [Azure Monitor Open Telemetry Exporter](https://www.nuget.org/packages/Azure.Monitor.OpenTelemetry.Exporter#readme-body-tab) package.  If you're looking for low-level instrumentation, the [Azure SDK logs](https://learn.microsoft.com/dotnet/azure/sdk/logging) emit data for every service operation and its result as well as client state.   As you appear familiar with those, I'll leave out the details.  

To summarize all of that:
  - There's really no need to monitor service traffic to understand processor health.  So long as it has been started, the processor is self-healing and will continue to attempt receives.   Your error handler will be notified of errors as they're encountered.   If your error handler is not seeing activity, the processor is not experiencing problems.
  
- To monitor activities related to processing, the OTel traces are the best approach and follow a well-documented standard.

- To watch individual service operations and client state, the Azure SDK logs that you're referring to are the only option.

**Comment by cjablonski76** [comment]
Ok, @jsquire, that all makes sense. I think what is throwing me then is that I keep seeing some exception bubble to my error handle method.
It appears that when our containers are being shutdown (as part of natural events), we're getting hit with exceptions like:

```
ErrorSource: Receive.
Exception: System.ObjectDisposedException:
The message processor is unable to continue and will stop processing;
the host application has closed the connection to the Service Bus service.
Object name: 'ServiceBusConnection'.
System.ObjectDisposedException:
The message processor is unable to continue and will stop processing; the host application has closed the connection to the Service Bus service.
Object name: 'ServiceBusConnection'.
```

I assume this isn't expected behavior, right?

**Comment by cjablonski76** [comment]
whoops, forgot to add the /unresolve

**Comment by jsquire** [comment]
@cjablonski76 :  The TLDR is that that is normal when you don't stop your processor.

That exception is raised when the processor isn't shut down gracefully and the `ServiceBusClient` that spawned the processor is closed/disposed while the processor is running.  Since the underlying connection is owned by the client, the processor loses access to the connection while still running.   Because this only happens when an application intentionally calls CloseAsync or DisposeAsync on the client (directly or via DI cleanup), the processor knows that it is not allowed to recover the connection and, instead, allows its internal receivers to crash.

You can prevent this by stopping the processor before cleaning up the client.

**Comment by cjablonski76** [comment]
> @cjablonski76 : The TLDR is that that is normal when you don't stop your processor.
> 
> That exception is raised when the processor isn't shut down gracefully and the `ServiceBusClient` that spawned the processor is closed/disposed while the processor is running. Since the underlying connection is owned by the client, the processor loses access to the connection while still running. Because this only happens when an application intentionally calls CloseAsync or DisposeAsync on the client (directly or via DI cleanup), the processor knows that it is not allowed to recover the connection and, instead, allows its internal receivers to crash.
> 
> You can prevent this by stopping the processor before cleaning up the client.

Ok, that's what I was afraid of. I'll have to figure out how a way to stop the processor gracefully when the app shuts down.
Thanks!
